src/proxy/pool.rs

```diff
@@ -12,80 +12,539 @@
+                .established_conn_writelock
+                .get(&hash_key, &guard)
+                .unwrap();
+            trace!("fallback attempt - got keyed lock out of lockmap");
+            exist_conn_lock.as_ref().unwrap().clone()
+        };
+
+        debug!(
+            "appears we need a new conn, attempting to win connlock for wl key {:#?}",
+            workload_key
+        );
+        let got_conn = match inner_conn_lock.try_lock() {
+            Ok(_guard) => {
+                // BEGIN take inner writelock
+                // If we get here, it means the following are true:
+                // 1. We did not get a connection for our key.
+                // 2. We have the exclusive inner writelock to create a new connection for our key.
+                //
+                // So, carry on doing that.
+                debug!("nothing else is creating a conn and we won the lock, make one");
+                let pool_conn = self.spawn_new_pool_conn(workload_key.clone()).await;
+                let client = ConnClient {
+                    sender: pool_conn?,
+                    stream_count: Arc::new(AtomicU16::new(0)),
+                    stream_count_max: self.max_streamcount,
+                    wl_key: workload_key.clone(),
+                };
+
+                debug!(
+                    "starting new conn for key {:#?} with pk {:#?}",
+                    workload_key, pool_key
+                );
+                Some(client)
+                // END take inner writelock
+            }
+            Err(_) => {
+                debug!("we didnt' win the lock, something else is creating a conn, wait for it");
+                // If we get here, it means the following are true:
+                // 1. At one point, there was a preexisting conn in the pool for this key.
+                // 2. When we checked, we got nothing for that key.
+                // 3. We could not get the exclusive inner writelock to add a new one for this key.
+                // 4. Someone else got the exclusive inner writelock, and is adding a new one for this key.
+                //
+                // So, loop and wait for the pool_watcher to tell us a new conn was enpooled,
+                // so we can pull it out and check it.
+                loop {
+                    match self.pool_watcher.changed().await {
+                        Ok(_) => {
+                            trace!(
+                                "notified a new conn was enpooled, checking for hash {:#?}",
+                                hash_key
+                            );
+                            // The sharded mutex for this connkey is already locked - someone else must be making a conn
+                            // if they are, try to wait for it, but bail if we find one and it's got a maxed streamcount.
+                            let existing_conn = self.state.connected_pool.get(&hash_key);
+                            match existing_conn {
```

Comment:
I don't understand this. 

This loop should get triggered by a pool_watcher event. pool_watcher can come from _anyone_ notifying (i.e. different keys).

Then we read from the pool immediately. This is not a mutex that is held for a long time (basically just to read and immediately unlock, not held across IO.

So the "try to wait for it" comment confuses me

Reply:
`wait` in the sense that we will wait for pool watcher to resolve, and if it resolves and we still don't have anything, loop again - not in the sense that there's a `sleep` here (though basically we are `await`ing the watcher which is functionally a `yield` from the runtime perspective)

Reply:
Reworded, this was an old, out of place comment anyway